lenskit-svd/src/main/java/org/lenskit/mf/BPR/BPRMFModelProvider.java

```diff
@@ -0,0 +1,175 @@
+                              @RegularizationTerm double regularizaiton,
+                              @Transient StoppingCondition stoppingCondition,
+                              @Transient Random rand) {
+        this.dao = dao;
+        this.pairGenerator = pairGenerator;
+        this.featureCount = featureCount;
+        this.learningRate = learningRate;
+        this.regularization = regularizaiton;
+        this.stoppingCondition = stoppingCondition;
+        this.rand = rand;
+
+        itemIndex = new HashKeyIndex();
+        userIndex = new HashKeyIndex();
+    }
+
+    @Override
+    public MFModel get() {
+        //REVIEW arbitrary size.
+        // This will accumulate BPR-Opt (minus the regularization) and will be negated to make an error.
+        // -30 is arbitrary, but would indicate a _really_ consistently bad prediction (~ p=1*10^-13),
+        // and is therefore a reasonable "max_error".
+        RollingWindowMeanAccumulator optAccum = new RollingWindowMeanAccumulator(1000, -30);
+
+        int userCount = dao.getEntityIds(CommonTypes.USER).size();
+        RealMatrix userFeatures = MatrixUtils.createRealMatrix(userCount, featureCount);
+        initFeatures(userFeatures);
+
+        int itemCount = dao.getEntityIds(CommonTypes.ITEM).size();
+        RealMatrix itemFeatures = MatrixUtils.createRealMatrix(itemCount, featureCount);
+        initFeatures(userFeatures);
+
+        logger.debug("Learning rate is {}", learningRate);
+        logger.debug("Regularization term is {}", regularization);
+
+        logger.info("Building MPR-MF with {} features for {} users and {} items",
+                featureCount, userCount, itemCount);
+
+        TrainingLoopController controller = stoppingCondition.newLoop();
+
+        //REVIEW: because of the nature of training samples (and the point that the BPR paper makes that training
+        // by-item or by-user are not optimal) one "iteration" here will be one training update. This leads to _really_
+        // big iteration counts, which can actually overflow ints!. one suggestion would be to allow the iteration count
+        // controller to accept longs, but I don't know if this will introduce backwards compatibility issues (I imagine
+        // this depends on the robustness of our type conversion in the configuration.
+        while(controller.keepTraining(-optAccum.getMean())) {
+            TrainingItemPair pair = pairGenerator.nextPair();
+            // Note: bad code style variable names are generally to match BPR paper and enable easier implementation
+            long iid = pair.g;
+            int i = itemIndex.internId(iid);
+            long jid = pair.l;
+            int j = itemIndex.internId(jid);
+            long uid = pair.u;
+            int u = userIndex.internId(uid);
+
+            RealVector w_u = userFeatures.getRowVector(u);
+            RealVector h_i = itemFeatures.getRowVector(i);
+            RealVector h_j = itemFeatures.getRowVector(j);
```

Comment:
[`Vectors.matrixRow`](http://lenskit.org/master/apidocs/org/lenskit/util/math/Vectors.html#matrixRow-org.apache.commons.math3.linear.RealMatrix-int-) will return a flyweight backed by the matrix; `getRowVector` will copy, I believe.